부산대 FE 손정훈 - 2단계 로그인, 주문하기 API 구현하기

SonJH7 opened the pull request:
## 구현한 내용
- 로그인 성공 시 사용자 정보 저장 후 전역 상태 갱신
- 주문 페이지에서 상품 요약 정보를 불러오고, 선택한 카드와 수신인 목록을 기반으로 주문 API 호출

## 질문할 내용
- OrderPage가 한 파일에 너무 많은 역할을 담고 있어 가독성이 떨어진다고 생각합니다.  스타일 정의를 별도 파일로 분리하거나 어떤 기준으로 하위 컴포넌트(예: 카드 선택, 메시지 입력, 수신인 관리)를 분리하면 좋을지 궁금합니다.

- API 호출 후 응답 검증에 실패하거나 네트워크 오류가 발생했을 때 토스트 메시지를 노출하도록 구현했는데, 이 로직을 전역적으로 처리할 방법(예: 에러 전용 훅이나 상태 관리 도구)을 고려하는게 좋을까요?

SonJH7 committed:
Update README.md

SonJH7 committed:
fea(login): login 요청을 보내는 api 함수 작성

SonJH7 committed:
feat(login): 로그인 요청을 보내는 api 함수 작성
- 응답에서 인증 토큰, 이메일 및 이름을 검증
-세션 저장 유틸리티로 이용자 정보 유지 및 제공(AuthContext)
- 로그인 오류 발생시 토스트 메시지 표시

SonJH7 committed:
fix(loginpage): 로그인 폼에서 직접 로그인 api 호출하도록 수정

- 로그인 페이지가 아닌 컴포넌트 단에서 로그인 api 호출하도록 수정

- loginpage의 콜백을 단순화

SonJH7 committed:
feat(productapi): 상품 요약 정보를 불러오는 api 호출 로직 작성

- 상품 요약 정보를 조회해 상태에 저장
- 요약 정보 바탕으로 화면에 표시

SonJH7 committed:
feat(apiorder): 주문 API 호출 함수 추가

- 서버에 주문 정보를 전송
- 로그인한 이용자의 이름을 기본 값으로 넣고, 주문 성공 시 홈으로 이동

SeonJun-Hwang commented on src/api/order.ts:
```diff
@@ -0,0 +1,36 @@
+  })
+
+  if (typeof data !== 'object' || typeof data.success !== 'boolean') {
+    throw new Error('Invalid response from /api/order')
```
해당 에러는 에러 요청 실패가 아니라 에러 내용에 대한 에러여야해요
url 요청에 대한 부분은 useFetchAPI 내부에서 공통적으로 처리되어야 합니다.

SonJH7 replied:
이번에도 정말 피가 되고 살이 되는 피드백 감사합니다. 덕분에 order 페이지에 대한 리팩토링을 하고 싶었는데 가이드가 절실히 필요한 상황에서 완벽한 가이드와 함께 어느 정도 리팩토링 욕구(?)를 해소한 것 같습니다! 감사합니다!

추가적으로 질문이 있습니다!
컴포넌트 폴더 안 파일이 많은데 그대로 두어도 괜찮을까요??
아니면 성격에 따라 분리하거나 하는 방식을 적용 하는 것이 좋을까요??

SeonJun-Hwang replied:
components 폴더로 묶은 이후에는 정답이 없긴 합니다
어떻게 묶어보니 편하더라 하는 경험치를 쌓아보셔도 될 듯 합니다.

SeonJun-Hwang reviewed (changes requested):
과제 구현간 고생 많으셨습니다 🎉 
로그인 정보를 context로 분리하고 필요한 곳에서 호출해 쓰는 방향으로 잘 분리하신거 같아요 💯 
실제로도 저런 식으로 많이 분리해서 사용합니다.

1. 이야기 해주신 OrderPage에서 보듯이 실제로 사용되는 컴포넌트 1개에 비해 너무 많은 내용이 들어있습니다. 먼저 같은 내용을 구성하지만, 역할이 다른 style 파일을 분리하는게 가장 먼저인거 같아요
그 다음으로 다른 컴포넌트간 공통으로 사용할 수 있는 부분을 공통화하는게 순서로 보입니다.
하위 컴포넌트 분리는 그 이후에 자연스럽게 보일듯 합니다.

2. 이부분도 custom hook으로 어느정도 분리가 가능합니다. Errorboundary 같은 내용도 있지만, 지금 적용하기에는 toast가 조금 더 적합해보입니다. 궁금하시다면 공부만 해보시는걸 추천드립니다. ( 이부분은 다음 과제에 언급이 되어있네요, 구현은 참아주세요 😢  )

OrderPage 관련해서 질문을 주신 만큼, Approve로 다음 과제를 개발하는 것보다, 코드를 분리해주신 후 어떤 관점에서 왜 분리했는지를 추가 리뷰에 알려주시면, 한번 더 보고 Apporove 드리는걸로 진행하겠습니다 😄

SonJH7 committed:
refactor(orderpage): 기존 페이지의 스타일 코드를 분리

SonJH7 committed:
refactor(orderpage): 공통으로 이용하는 에러 메시지와 노란색 버튼 스타일을 common.ts에 정의

- OrderPage.styles.ts에서 공통 스타일을 가져와 재이용하고 주문 버튼도 공통 버튼 기반 수정
- 로그인 폼과  공통 스타일을 가져와 재이용하고 주문 버튼도 공통 버튼 기반으로 수정

- 받는 사람 모달도 공통버튼과 에러 메시지 컴포넌트 활용하도록 수정

SonJH7 committed:
refactor(orderpage): 하위 컴포넌트 분리

- 카드 선택 목록을 별도 컴포넌트로 분리해 간결하게 코드 작성

- 받는 목록 표시도 분리하여 코드 중복 감소

SonJH7 committed:
fix(orderapi): 중복된 응답 검증을 제거 및 실 패 시 메시지 전달 하도록 수정

SeonJun-Hwang reviewed (approved):
리팩토링까지 고생하셨습니다 🎉 
협업에 가장 먼저 고민하는 부분이 폴더 구조를 어떻게 짜둬야 나중에 보기 편할까 였던거 같아요
적용까지 고생 많으셨습니다!
